Initial Import to implement once creation procedures.

jvelilla opened the pull request:
Updated code generation to generate factory funtions when we have
a once procedure as a creation procedure.

Added a validity check {EIFFEL_CLASS_C} to ensure once creation procedures does not
have arguments

jvelilla committed:
Initial Import to implement once creation procedures.

Updated code generation to generate factory funtions when we have
a once procedure as a creation procedure.

Added a validity check {EIFFEL_CLASS_C} to ensure once creation procedures does not
have arguments

jvelilla committed:
Merge branch 'master' of https://github.com/EiffelSoftware/EiffelStudio into es_enums_v2

kwaxer commented on Src/Eiffel/eiffel/interface/eiffel_class_c.e:
```diff
@@ -1909,6 +1914,56 @@ feature -- Supplier checking
 
+feature -- Once Creation Procedure
+
+	check_once_creation_procedure (ast_b: CLASS_AS)
```
This code does not work if an inherited once feature is used as a creation procedure.

So, the validity check should be done somewhere else.

Also, the validity rule applies only to once classes, there is no restriction on arguments for once creation procedures in non-once classes.

jvelilla replied:
Any suggestion as to where we should do the validity check?

kwaxer commented on Src/Eiffel/eiffel/interface/feature_i.e:
```diff
@@ -3443,6 +3443,105 @@ feature -- C code generation
+		do
+			if used then
+				tmp_body_index := body_index
+				l_byte_code := tmp_opt_byte_server.disk_item (tmp_body_index)
```
Given that the byte code is generated from scratch, it does not seem reasonable to load it and modify immediately (moreover, it may break code generation for different class types). New BYTE_CODE should be used instead.

kwaxer commented on Src/Eiffel/eiffel/interface/feature_i.e:
```diff
@@ -3443,6 +3443,105 @@ feature -- C code generation
+				end
+
+				create l_assign
+				l_assign.set_target (generate_target_access_factory)
```
Shouldn't it simply be `...set_target (create {RESULT_B})`?

kwaxer commented on Src/Eiffel/eiffel/interface/feature_i.e:
```diff
@@ -3443,6 +3443,105 @@ feature -- C code generation
+					l_byte_context.clear_feature_data
+				end
+			else
+				System.removed_log_file.add (class_type, feature_name)
```
This has been reported already for a creation procedure itself.

kwaxer commented on Src/Eiffel/eiffel/interface/feature_i.e:
```diff
@@ -3443,6 +3443,105 @@ feature -- C code generation
+		do
+			create l_result
+			l_result.set_info (create {CREATE_TYPE}.make (a_class_type.associated_class.actual_type))
+			l_result.set_is_active (True)
```
This requires discussion and clear definition in the language standard, otherwise the behavior is unclear.

kwaxer commented on Src/Eiffel/eiffel/interface/feature_i.e:
```diff
@@ -3443,6 +3443,105 @@ feature -- C code generation
+			l_info: CREATE_INFO
+			l_feature: FEATURE_B
+		do
+			create l_result
```
Why not simply change function result to `CREATION_EXPR_B`?

kwaxer commented on Src/Eiffel/eiffel/interface/feature_i.e:
```diff
@@ -3443,6 +3443,105 @@ feature -- C code generation
+					-- the associated class of the current type.
+				l_byte_context := byte_context
+
+				if System.in_final_mode and then System.inlining_on then
```
It is better to avoid copy-paste (i.e. code duplication) when possible.

jvelilla committed:
Commented code until we find the correct place to validate that once creation
procedures can't have arguments, it should only be applied to once classes.

Improve funtion factory for once creation procedure. BYTE_CODE generated from
scracth added support to generate factory funtion with arguments.

jvelilla committed:
Merge branch 'master' of https://github.com/EiffelSoftware/EiffelStudio into es_enums_v2

jvelilla committed:
Added Validity rules for once creation procedures at DEGREE_3
*It must be immediate (i.e. not inherited).
*It may not be used in a Call (qualified or not) (i.e., only in a creation all).
*The class is not generic.

jvelilla committed:
Updated cde to add validity error messages for once creation procedures

jvelilla committed:
Added missing classes VGCC9 and VGCC10

jvelilla committed:
Updated VGCC9 and VGCC10 to make the messages translatable.
Updated Once creation generation.

jvelilla committed:
Merge branch 'master' of https://github.com/EiffelSoftware/EiffelStudio into es_enums_v2

jvelilla committed:
Removed unneeded check, now once creation procedures support can have arguments.
Added a feature to FEATURE_B class to mark a current feature as instance_free.
Update call for once creation procedures, we use them as instance free calls.
Updated CREATION_EXPR_B as a FEATURE_B iff it's a once creation procedure.
